### ISSUE TITLE
How to use "dynamic queries" with Dynamic columns #964 feature?

### ISSUE THREAD

**Issue by drekoo** [linq2db/linq2db#1083]
Hi,
it seems that 
```c#
var filteringColumn = "FirstName";
var sortingColumn = "LastName";
var filter = "John";

db.GetTable<Person>()
    .Where(p => Sql.Property<object>(p, filteringColumn) == filter)
    .OrderBy(p => Sql.Property<object>(p, sortingColumn))
    .ToList();
```
from current #964 feature implementation is actualy not "fully implemented". "Only constant strings are alowed for member name in Sql.Property expressions.". **Will this be implemented in the future**.

I have entity

```c#
[Table("PRODUCT")]
public partial class Product
{
...
 public double PRICE1 { get; set; }
 public double PRICE2 { get; set; }
 public double PRICE3 { get; set; }
 public double PRICE4 { get; set; }
 public double PRICE5 { get; set; }
...
}

//I want to query with

string whichPriceName = "PRICE" + settings.PriceID.ToString();

from p in dbContext.Db.GetTable<Product>()
                       select new ProductDTO
                       {
                           ID = p.ID,
                           PRICE = Sql.Property<double>(p, whichPriceName )
                       };

//Possible (current) solution to my problem is using expression method in Product Entity

 [ExpressionMethod("WhichPriceExpression")]
        public double WhichPrice(int priceID)
        {
            //local
            switch (priceID)
            {
                case 5:
                    return PRICE5;
                case 4:
                    return PRICE4;
                case 3:
                    return PRICE3;
                case 2:
                    return PRICE2;
                default:
                    return PRICE1;
            }

        }

        public static Expression<Func<Product, int, double>> WhichPriceExpression()
        {
            return (product, index) => index == 5 ?
            product.PRICE5: index == 4 ?
            product.PRICE4: index == 3 ?
            product.PRICE3: index == 2 ?
            product.PRICE2: product.PRICE1;
        }

          from p in dbContext.Db.GetTable<Product>()
                       select new ProductDTO
                       {
                           ID = p.ID,
                           PRICE = p.WhichPrice(settings.PriceID)
                       };

```
The expression method is actually translated to SQL CASE statement which is not nice, if there are a lot of PRICE{X} fields...

**Is technically possible in Linq2DB to have expression method with parameter....**

```c#

 public static Expression<Func<Product, double>> WhichPriceExpression(int priceID)
        {
            switch (priceID)
            {
                case 5:
                    return (product) => product.PRICE5;
                case 4:
                    return (product) => product.PRICE4;
                case 3:
                    return (product) => product.PRICE3;
                case 2:
                    return (product) => product.PRICE2;
                default:
                    return (product) => product.PRICE1;
            }
        }
```

**Comment by MaceWindu** [comment]
I will take a look how to add support for it

**Comment by sdanyliv** [comment]
Unfortunately yo can not use this feature in that way. I will extend it. But for your case i have created working solution using `Sql.Extension`:
```cs
public static class MySql
{
	class WithPriceBuilder : Sql.IExtensionCallBuilder
	{
		public void Build(Sql.ISqExtensionBuilder builder)
		{
			var index = builder.GetValue<int>(1);
			var field = builder.GetExpression(0) as SqlField;
			if (field == null)
				throw new Exception("Entity required as paramater");

			var fieldName = "PRICE" + index;
			var table = (SqlTable)field.Table;
			if (!table.Fields.TryGetValue(fieldName, out var newField))
			{
				newField = new SqlField();
				newField.Name = fieldName;
				newField.Table = table;
				table.Fields.Add(fieldName, newField);
			}

			builder.ResultExpression = newField;
		}
	}

	[Sql.Extension("", BuilderType = typeof(WithPriceBuilder), ServerSideOnly = true)]
	public static double WithPrice<T>(T entity, [SqlQueryDependent] int index)
	{
		throw new NotImplementedException();
	}
}
```
usage
```cs
var whichPriceName = 1;
var query = from p in db.GetTable<Product>()
	select new ProductDTO
	{
		ID = p.PriceID,
		PRICE = MySql.WithPrice(p, whichPriceName)
	};
```
Not so trivial but before i fix Dynamic Columns it's only one way.

**Comment by drekoo** [comment]
Thanks

**Comment by drekoo** [comment]
Just for info, I had to change code, because in my example I 'm using grouping and expression is not SqlField but SqlColumn.  This is working, but maybe just for one (nested) level.... 

```
public void Build(Sql.ISqExtensionBuilder builder)
            {
                var fieldName = builder.GetValue<string>(1);
                var fieldExpr = builder.GetExpression(0);

                SqlField field = null;

                if (fieldExpr is SqlColumn)
                    field = (fieldExpr as SqlColumn).Expression as SqlField;
                else
                    field = builder.GetExpression(0) as SqlField;

                if (field == null)
                    throw new Exception("Entity required as parameter");

                var table = (SqlTable)field.Table;
                if (!table.Fields.TryGetValue(fieldName, out var newField))
                {
                    newField = new SqlField
                    {
                        Name = fieldName,
                        Table = table
                    };
                    table.Fields.Add(fieldName, newField);
                }

                builder.ResultExpression = newField;
            }
```

**Comment by sdanyliv** [comment]
Show me your linq query

**Comment by drekoo** [comment]
It is quite complex , Can't show it all because of production code, but the main "point" is here:
...I have made MySql.SqlColumn(a, PriceFieldName) extension...
See 
Price = MySql.SqlColumn(a, PriceFieldName) bellow

```
...
 qProducts = from a in dbContext.Db.GetTable<Product>()
                       join z in qStock on a.ID equals z.ID
                       from ng in dbContext.Db.GetTable<Group>().LeftJoin(j => j.ID == a.Group)
                       where
                       isFilterEmpty || (filter.Groups.Contains(a.Group) || filter.ParentGroup.Contains(ng.ParentGroup) || filter.CommodityGroup.Contains(a.CommodityGroup))
                       group z by new { a.ID, a.Name, a.Name2, a.Unit, a.Group, ng.ParentGroup, a.CommodityGroup, Price = MySql.SqlColumn(a, PriceFieldName), a.Vat, a.Discount, a.ExtraData } into g
                       select new ProductDTO
                       {
                           ID= g.Key.ID,
                           Name = g.Key.Name ,
                           Name2 = g.Key.Name2 ,
                           Unit= g.Key.Unit,
                           Group= g.Key.Group,
                           CommodityGroup = g.Key.CommodityGroup,
                           ParentGroup= g.Key.ParentGroup,
                           Price = g.Key.Price.GetValueOrDefault(),
                           Discount = g.Key.Discount.GetValueOrDefault(),
                           Vat= g.Key.Vat,
                           ExtraData = g.Key.ExtraData,
                           Stock = g.Sum(s => s.Quantity).GetValueOrDefault()
                       };
...
```
After that I have Where filters....

**Comment by MaceWindu** [comment]
should be fixed by #1102
